File: benchmarks/ptb/bench_bot.py
"""PTB benchmark bot — identical features to Rust versions."""
import asyncio
import os
import logging

import uvicorn
from starlette.applications import Starlette
from starlette.requests import Request
from starlette.responses import PlainTextResponse, Response
from starlette.routing import Route

from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.constants import ChatAction
from telegram.ext import (
    Application,
    CommandHandler,
    CallbackQueryHandler,
    MessageHandler,
    ContextTypes,
    filters,
)

logging.basicConfig(level=logging.WARNING)

TOKEN = os.environ["TELEGRAM_BOT_TOKEN"]
URL = os.environ["WEBHOOK_URL"]


async def start(update: Update, context: ContextTypes.DEFAULT_TYPE):
    name = update.effective_user.first_name if update.effective_user else "there"
    keyboard = [
        [
            InlineKeyboardButton("Option 1", callback_data="1"),
            InlineKeyboardButton("Option 2", callback_data="2"),
        ],
        [InlineKeyboardButton("Option 3", callback_data="3")],
    ]
    await update.message.reply_text(
        f"Hi {name}! I am a benchmark bot.\nUse /help for info.",
        reply_markup=InlineKeyboardMarkup(keyboard),
    )


async def help_cmd(update: Update, context: ContextTypes.DEFAULT_TYPE):
    await update.message.reply_text(
        "Commands: /start, /help\n"
        "Send any text to echo.\n"
        "Press inline buttons to test callbacks."
    )


async def echo(update: Update, context: ContextTypes.DEFAULT_TYPE):
    await context.bot.send_chat_action(
        chat_id=update.effective_chat.id, action=ChatAction.TYPING
    )
    await update.message.reply_text(update.message.text)


async def button(update: Update, context: ContextTypes.DEFAULT_TYPE):
    query = update.callback_query
    await query.answer()
    await query.edit_message_text(text=f"You selected: Option {query.data}")


async def main():
    app = Application.builder().token(TOKEN).updater(None).build()
    app.add_handler(CommandHandler("start", start))
    app.add_handler(CommandHandler("help", help_cmd))
    app.add_handler(CallbackQueryHandler(button))
    app.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, echo))

    await app.bot.set_webhook(
        url=f"{URL}/telegram", allowed_updates=Update.ALL_TYPES
    )

    async def telegram(request: Request):
        data = await request.json()
        await app.update_queue.put(
            Update.de_json(data=data, bot=app.bot)
        )
        return Response()

    async def health(_: Request):
        return PlainTextResponse("OK")

    starlette_app = Starlette(
        routes=[
            Route("/telegram", telegram, methods=["POST"]),
            Route("/healthcheck", health, methods=["GET"]),
        ]
    )
    server = uvicorn.Server(
        config=uvicorn.Config(
            app=starlette_app,
            port=8000,
            host="127.0.0.1",
            use_colors=False,
            log_level="warning",
        )
    )

    async with app:
        await app.start()
        print("PTB benchmark bot running on port 8000. Send /start to test.")
        await server.serve()
        await app.stop()


if __name__ == "__main__":
    asyncio.run(main())

File: crates/telegram-bot/examples/bench_bot.rs
//! RTB benchmark bot -- identical features to PTB and teloxide versions.
//!
//! Features: /start (with inline keyboard), /help, echo with typing action,
//! callback query handler, webhook on port 8000 via `run_webhook()`.
//!
//! # Usage
//!
//! ```sh
//! TELEGRAM_BOT_TOKEN="your-token" \
//! WEBHOOK_URL="https://your.domain" \
//! cargo run -p rust-tg-bot --example bench_bot --features webhooks
//! ```

use std::sync::Arc;

use rust_tg_bot::ext::prelude::{
    ApplicationBuilder, CommandHandler, Context, FnHandler, HandlerResult, InlineKeyboardButton,
    InlineKeyboardMarkup, MessageHandler, Update, COMMAND, TEXT,
};
use rust_tg_bot::ext::updater::WebhookConfig;
use rust_tg_bot::raw::bot::ChatId;

// -- Handlers ----------------------------------------------------------------

async fn start(update: Arc<Update>, context: Context) -> HandlerResult {
    let name = update
        .effective_user()
        .map(|u| u.first_name.as_str())
        .unwrap_or("there");
    let chat_id = update.effective_chat().map(|c| c.id).unwrap_or(0);

    let keyboard = InlineKeyboardMarkup::new(vec![
        vec![
            InlineKeyboardButton::callback("Option 1", "1"),
            InlineKeyboardButton::callback("Option 2", "2"),
        ],
        vec![InlineKeyboardButton::callback("Option 3", "3")],
    ]);

    context
        .bot()
        .send_message(
            chat_id,
            &format!("Hi {name}! I am a benchmark bot.\nUse /help for info."),
        )
        .reply_markup(serde_json::to_value(&keyboard).unwrap_or_default())
        .await?;
    Ok(())
}

async fn help_cmd(update: Arc<Update>, context: Context) -> HandlerResult {
    context
        .reply_text(
            &update,
            "Commands: /start, /help\nSend any text to echo.\nPress inline buttons to test callbacks.",
        )
        .await?;
    Ok(())
}

async fn echo(update: Arc<Update>, context: Context) -> HandlerResult {
    let msg = match update.effective_message() {
        Some(m) => m,
        None => return Ok(()),
    };
    let chat_id = msg.chat.id;
    let text = msg.text.as_deref().unwrap_or("");
    if text.is_empty() {
        return Ok(());
    }

    context
        .bot()
        .send_chat_action(ChatId::Id(chat_id), "typing")
        .await
        .ok();
    context.bot().send_message(chat_id, text).await?;
    Ok(())
}

async fn button_callback(update: Arc<Update>, context: Context) -> HandlerResult {
    let cq = match update.callback_query() {
        Some(c) => c,
        None => return Ok(()),
    };
    let data = cq.data.as_deref().unwrap_or("?");
    context.bot().answer_callback_query(&cq.id).await?;
    if let Some(msg) = cq.message.as_ref().and_then(|m| m.as_message()) {
        context
            .bot()
            .edit_message_text(&format!("You selected: Option {data}"))
            .chat_id(msg.chat.id)
            .message_id(msg.message_id)
            .await?;
    }
    Ok(())
}

// -- Main --------------------------------------------------------------------

#[tokio::main]
async fn main() {
    tracing_subscriber::fmt()
        .with_max_level(tracing::Level::WARN)
        .init();

    let token = std::env::var("TELEGRAM_BOT_TOKEN").expect("TELEGRAM_BOT_TOKEN required");
    let webhook_url = std::env::var("WEBHOOK_URL").expect("WEBHOOK_URL required");

    let app = ApplicationBuilder::new().token(&token).build();

    app.add_handler(CommandHandler::new("start", start), 0)
        .await;
    app.add_handler(CommandHandler::new("help", help_cmd), 0)
        .await;
    app.add_handler(FnHandler::on_callback_query(button_callback), 0)
        .await;
    app.add_handler(MessageHandler::new(TEXT() & !COMMAND(), echo), 0)
        .await;

    println!("RTB benchmark bot starting on port 8000. Send /start to test.");

    let config = WebhookConfig::new(format!("{webhook_url}/telegram"))
        .port(8000)
        .url_path("/telegram")
        .secret_token("bench-secret-token");

    if let Err(e) = app.run_webhook(config).await {
        eprintln!("Error: {e}");
    }
}
